Call private cache methods only for `OC\Files\Cache\Cache`

d--j opened the pull request:
The two methods `correctFolderSize` and `calculateFolderSize` should be
part of the interface since they are used  throughout the code when
an instance of this interface is expected. The implementation
`OC\Files\Cache\FailedCache` was missing this methods.

This fixes #23534

icewind1991 commented:
They were left out on purpose because they are specific to our internal cache implementation.

Instead of adding them to the interface (requiring all 3rdparty implementations to implement them) the code that calls  these methods should do an `instanceof OC\Files\Cache\Cache`

DeepDiver1975 commented:
@icewind1991 please comment on this

I tend to keep them hidden since they are mote on the nature of implementation details - THX

d--j commented:
OK, but they are missing in `OC\Files\Cache\FailedCache`, aren't they?

DeepDiver1975 commented:
THX for the quick response @icewind1991

icewind1991 commented:
`FailedCache` doesn't need them, and if the `instanceof` check is in place it wont try to call it

nickvergessen commented on lib/private/files/cache/watcher.php:
```diff
@@ -106,7 +106,9 @@ public function update($path, $cachedData) {
 			$this->cleanFolder($path);
 		}
-		$this->cache->correctFolderSize($path);
+		if ($this->cache instanceof Cache) {
```
this should fail for cache wrappers?

icewind1991 replied:
`CacheWrapper` extends `Cache` so it should be fine

d--j commented:
@owncloud-bot: My contribution is MIT licensed. (I also sent in a CLA)

icewind1991 commented on lib/private/files/cache/updater.php:
```diff
@@ -187,8 +191,12 @@ public function renameFromStorage(IStorage $sourceStorage, $source, $target) {
+		if ($sourceCache instanceof Cache) {
+			$sourceCache->correctFolderSize($source);
+		}
+		if ($this->cache instanceof Cache) {
```
You can combine these

d--j replied:
@icewind1991 Do you mean
```php
    if ($sourceCache instanceof Cache && $this->cache instanceof Cache) {
      $sourceCache->correctFolderSize($source);
      $this->cache->correctFolderSize($target);
    }
```
with combining?

icewind1991 replied:
Nevermind, I missed the difference between `$sourceCache` and `$this->cache`, it's good as is

icewind1991 commented:
Looks good otherwise

icewind1991 commented:
:+1:

d--j committed:
Call private cache methods only for `OC\Files\Cache\Cache`

The two implementation detail methods `correctFolderSize` and
`calculateFolderSize` should only be called for instances of
`OC\Files\Cache\Cache`. This commit adds guarding checks
whenever they are called.

d--j commented:
I changed ` OC\Files\Cache\Update` as @icewind1991 suggested and rebased to current master.

d--j commented:
@icewind1991 sorry I missed your line comment

> Nevermind, I missed the difference between $sourceCache and $this->cache, it's good as is

I just reverted the combining

PVince81 commented:
@icewind1991 please re-review the last changes

PVince81 commented:
PR for Jenkins/CI to see test results: https://github.com/owncloud/core/pull/24179

PVince81 commented:
@karlitschek backport to 9.0.2 ?

karlitschek commented:
please backport

icewind1991 commented:
:+1:

d--j commented:
For backporting: This version does not cleanly apply to the 9.0 branch anymore. Here is the same patch that should cleanly apply for the 9.0 branch: https://github.com/owncloud/core/files/222684/23972.patch.txt

PVince81 commented:
@d--j for the backport, we usually make separate PR, branching from "stable9" and that has "stable9" as target. Let me know if you want to do this yourself.

I'm still waiting for CI to finish here: https://github.com/owncloud/core/pull/24179

PVince81 commented:
Hmmm, in what situation is `$this->cache` not an instance of `Cache` ?

If it's the `FailedCache`, then why is it even going through all the scanner logic ? Can we short-circuit this ?

@icewind1991 mind explaining why you think this is the best solution ?

PVince81 commented:
After re-reading the description I think this is fine. If those are private methods then we better check the type.

PVince81 commented:
:+1:

DeepDiver1975 commented:
conflicting .... will take care
